providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java

```diff
@@ -247,53 +264,49 @@ private <T> ProviderEvaluation<T> getEvaluation(
-                .build();
+                    .errorCode(errorCode)
+                    .reason(errorCode.name())
+                    .value(defaultValue)
+                    .build();
         }
-        ProviderEvaluation<T> res;
+
         GoFeatureFlagUser user = GoFeatureFlagUser.fromEvaluationContext(evaluationContext);
         if (cache == null) {
+            // Cache is disabled we return directly the remote evaluation
             EvaluationResponse<T> proxyRes = resolveEvaluationGoFeatureFlagProxy(key, defaultValue, user, expectedType);
-            res = proxyRes.getProviderEvaluation();
-        } else {
-            res = getProviderEvaluationWithCheckCache(key, defaultValue, expectedType, user);
-            eventsPublisher.add(Event.builder()
-                .key(key)
-                .defaultValue(defaultValue)
-                .variation(res.getVariant())
-                .value(res.getValue())
-                .userKey(user.getKey())
-                .creationDate(System.currentTimeMillis())
-                .build()
-            );
+            return proxyRes.getProviderEvaluation();
         }
-        return res;
-    }
 
-    private <T> ProviderEvaluation getProviderEvaluationWithCheckCache(
-            String key, T defaultValue, Class<?> expectedType, GoFeatureFlagUser user) {
-        ProviderEvaluation<?> res;
+        String cacheKey = null;
         try {
-            String cacheKey = buildCacheKey(key, BeanUtils.buildKey(user));
-            res = cache.getIfPresent(cacheKey);
-            if (res == null) {
-                EvaluationResponse<T> proxyRes = resolveEvaluationGoFeatureFlagProxy(
-                    key, defaultValue, user, expectedType);
-                if (Boolean.TRUE.equals(proxyRes.getCachable())) {
-                    cache.put(cacheKey, proxyRes.getProviderEvaluation());
-                }
-                res = proxyRes.getProviderEvaluation();
-            } else {
-                res.setReason(CACHED_REASON);
+            cacheKey = buildCacheKey(key, BeanUtils.buildKey(user));
+            ProviderEvaluation<?> cacheResponse = cache.getIfPresent(cacheKey);
+            if (cacheResponse == null) {
+                throw new NotPresentInCache(cacheKey);
+            }
+            cacheResponse.setReason(CACHED_REASON);
+            eventsPublisher.add(Event.builder()
+                    .key(key)
+                    .kind("feature")
+                    .contextKind(user.isAnonymous() ? "anonymousUser" : "user")
+                    .defaultValue(defaultValue)
+                    .variation(cacheResponse.getVariant())
+                    .value(cacheResponse.getValue())
+                    .userKey(user.getKey())
+                    .creationDate(System.currentTimeMillis())
+                    .build()
+            );
+            return (ProviderEvaluation<T>) cacheResponse;
+        } catch (Exception e) {
+            if (!(e instanceof NotPresentInCache)) {
```

Comment:
it makes normal positive flow of first evaluation to get to exception - feels less natural and efficient, don't you think ?

Reply:
I don't really see the problem here, an `Exception` does not mean an error every time.

Reply:
yes that’s a good point, what I would like is to have all the flow in a single method and avoid as much as possible to duplicate code.

But I can have a look after the week end if you want